pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py

```diff
@@ -218,6 +223,71 @@ def _request_generator_thread(self):
             _LOGGER.debug('Sending request:\n%r', request)
             yield request
 
+    def _stop_request_generator(self, request_generator):
+        """Ensure a request generator is closed.
+
+        This **must** be done when recovering from a retry-able exception.
+        If not, then an inactive request generator (i.e. not attached to any
+        actual RPC) will be trying to access the same request queue as the
+        active request generator.
+
+        In addition, we want the gRPC thread consuming to cleanly exit so
+        that system resources are not wasted.
+
+        Args:
+            request_generator (Generator): A streaming pull request generator
+                returned from :meth:`_request_generator_thread`.
+
+        Returns:
+            bool: Indicates if the generator was successfully stopped. Will
+            be :data:`True` unless the queue is not empty and the generator
+            is running.
+        """
+        with self._put_lock:
+            try:
+                request_generator.close()
+            except ValueError:
+                # Should be ``ValueError('generator already executing')``
+                if not self._request_queue.empty():
+                    # This case may be a false negative in **very** rare
+                    # cases. We **assume** that the generator can't be
+                    # ``close()``-ed because it is blocking on ``get()``.
+                    # It's **very unlikely** that the generator was not
+                    # blocking, but instead **in between** the blocking
+                    # ``get()`` and the next ``yield`` / ``break``. However,
+                    # for practical purposes, we only need to stop the request
+                    # generator if the connection has timed out due to
+                    # inactivity, which indicates an empty queue.
+                    _LOGGER.debug(
+                        'Request generator could not be closed but '
+                        'request queue is not empty.')
+                    return False
+                # If we **cannot** close the request generator,
+                # then there is no blocking get on the queue. Since
```

Comment:
Do you mean that there *is* a blocking get?

Reply:
Yes, thanks.

Reply:
Updated